File: 1-100/2.add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int nCarry = 0;
        ListNode* pAnsNode = new ListNode(0);
        ListNode* pCurNode = pAnsNode;
        while (l1 != nullptr || l2 != nullptr)
        {
            int x = 0, y = 0;
            if (l1 != nullptr)
            {
                x = l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr)
            {
                y = l2->val;
                l2 = l2->next;
            }

            int sum = nCarry + x + y;
            nCarry = sum / 10;
            pCurNode->next = new ListNode(sum % 10);
            pCurNode = pCurNode->next;
        }

        if (nCarry > 0)
        {
            pCurNode->next = new ListNode(nCarry);
        }

        return pAnsNode->next;
    }
};

File: 1-100/2.add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        ret = ListNode(0)
        cur = ret
        carry = 0
        while l1 or l2:
            x, l1 = (l1.val, l1.next) if l1 else (0, None)
            y, l2 = (l2.val, l2.next) if l2 else (0, None)
            sum = carry + x + y
            carry, sum = (1, sum - 10) if sum >= 10 else (0, sum)
            cur.next = ListNode(sum)
            cur = cur.next
        
        if carry == 1:
            cur.next = ListNode(1)
        
        return ret.next
